#18 1) adding audio transcription

ndrean opened the pull request:
This is the first push where I add audio-to-text. I did not made any tests so far. The idea is more to check if the Readme is understandable and the code meaningful.

If you are still interested, I will continue with the second step:
- modify the DB to add the HNSWLib index.
- add embeddings
- add HNSWLib and the knn-search

LuchoTurtle commented:
Thank you so much for the PR <3
If you want me to take a look at it, even if it's not ready and for feedback, you can ping me :)

ndrean commented:
@luchoturtle

I try to further clean the code and Readme. I definitely need your eye on all this!

I still have a stupid error in the UI: I display an icon and a text "record" as the action button to record the audio. When you click it, I switch the CSS classes, but when I click to stop the recording, then the icon disappears. You maybe more clever than me!

LuchoTurtle commented:
Thank you for the work @ndrean! I'll take a look at it tomorrow! ❤️

ndrean commented:
Then the tests.....😥

ndrean commented:
ok, I found my mistake and corrected it. No more pb with the audio recording button!

LuchoTurtle commented:
Lovely, thanks a ton @ndrean !
I'm taking longer than usual with other stuff but I'll try to get to this today, even if it's later at night. Sorry about that! Can't wait 😍

ndrean commented:
The most interesting part is part 2!

I should cite my sources: 

1) <https://dockyard.com/blog/2023/03/07/audio-speech-recognition-in-elixir-with-whisper-bumblebee?utm_source=elixir-merge>
2) <https://dockyard.com/blog/2023/01/11/semantic-search-with-phoenix-axon-bumblebee-and-exfaiss>

LuchoTurtle commented:
I've made changes to the README. I kept most of the info, it was simply formatting and some writing style changes.
I've also added comments to `micro.js`. 

I'm now trying to get this to work but since the `elixir` version was changed in the PR, I'm trying to install `v16`.

LuchoTurtle commented on README.md:
Looks great :D 

Made some changes in [`b92221c` (#43)](https://github.com/dwyl/image-classifier/pull/43/commits/b92221cea44c525c204e8462ce0fbb6d60001479). But it's awesome, thanks!

LuchoTurtle commented on lib/app/speech_to_text.ex:
This makes sense :D 

However, inside `lib/app/models.ex`, we have a module that is testable and manages all the models that need to be downloaded. I think it's best to move the `serving` function from the `whisper` model to this manager and integrate it there so we can easily test it like we're already doing in the tests.

I think this is the best course of actions to get some unit tests going at a later stage.

ndrean replied:
Yes, totally

LuchoTurtle reviewed (changes requested):
I've tried it and the recording and audio transcription seems to be working smoothly! 
Thank you!

I'm marking this to "request some changes" but I'll gladly do the.
The only thing missing (besides the feedback I've given) are the tests. I can take care of both, if you want to :)

ndrean commented:
> since the elixir version was changed in the PR, I'm trying to install v16.

Ah yes, my version updated via brew, so I had to update the code. It is not disruptive (besides the order of the arguments in `File.stream` I believe; [source](https://hexdocs.pm/elixir/changelog.html).

ndrean commented:
@LuchoTurtle  In "models.ex", you have a guard Dialyzer warning when you use `info = if (model.load_featurizer) do...` because the first argument of the `if` macro is not a boolean.

Instead, you should use `Map.get/3`. The verbose explanation is because it can return `nil`, and [if](https://hexdocs.pm/elixir/case-cond-and-if.html#if-unless) executes the `else` part if the condition is false or nil so you can omit the `!= nil` condition.

```elixir
info = 
  if Map.get(model, :local_featurizer)  do
    {:ok, featurizer} = Bumblebee.load_featurizer(loading_settings)
    Map.put(info, :featurizer, featurizer)
  else
    info
  end
```

In my next code update, I modified this if you agree.

ndrean commented:
Check the last push: I modified Application.Ex and Models to include Whisper.
I basically added an argument to `App.Models.verify_and_download_models`. I also renamed the image model to `captioning_prod_model` (and the test to `captioning_test_model`). I hope this is inline with your thoughts.

LuchoTurtle commented:
Thanks for the changes :)

I'm trying to run `mix c` but I can't seem to run the tests, I get the following error:

```
** (Mix) Could not start application app: exited in: App.Application.start(:normal, [])
    ** (EXIT) an exception was raised:
        ** (MatchError) no match of right hand side value: {:error, "could not find file in local cache and outgoing traffic is disabled, url: https://huggingface.co/api/models/openai/whisper-small/tree/main"}
            (app 0.1.0) lib/app/models.ex:159: App.Models.load_offline_model/1
            (app 0.1.0) lib/app/models.ex:119: App.Models.whisper_serving/0
            (app 0.1.0) lib/app/application.ex:52: App.Application.start/2
            (kernel 9.2) application_master.erl:293: :application_master.start_it_old/4
```

Weird stuff, since the model is clearly downloaded :/

LuchoTurtle commented:
Running `mix phx.server` still works though :)

LuchoTurtle commented on lib/app/models.ex:
While I agree with some changes made to `models.ex`, it seems there is some duplication with explicitly naming the models in `application.ex`.

The reason `models.ex` existed was to have **everything** model-related inside this module so it would be easier to figure out what was going on. But with these changes, some bugs have been introduced and it kind of violates SRP. 

For example, the model already had managed test and prod models and downloaded accordingly depending on the environment. 

But now, in `application.ex`, you have:

```elixir
    [
      @whisper_model,
      @captioning_prod_model,
      @captioning_test_model
    ]
    |> Enum.each(&App.Models.verify_and_download_models/1)
```

This defeats the purpose of the original `verify_and_download_models/1`, which conditionally downloaded these models according to the env (if it was `:test` or `:prod`).  This was done through the `serving` functions.

I think this is more confusing now. I'll make some changes to this but keep the `Whisper` ones you've implemented in the `models.ex` file.

LuchoTurtle replied:
Fixed in [`8117b96` (#43)](https://github.com/dwyl/image-classifier/pull/43/commits/8117b9698dfa72fde623507659e0e29e257bf167). 

I've kept some changes you've made in `models.ex`, thank you!

I'll do the tests now :)

LuchoTurtle reviewed (approved):
@ndrean I've made changes:
- changed the `models.ex` and put all model-related code there to maintain consistency and single responsibility.
- added some notes on the README regarding `ffmpeg` and dividing some texts you've wrote into chapters (just to be more clear to the person reading).
- added a test to get coverage back to 100%.

I think this looks good for now but I'll wait for your approval to see if everything is cool to later work on the second PR #45  you've awesomely implemented :D

ndrean commented:
Much better indeed

LuchoTurtle commented:
I give the green light 🟢 😃
